lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -772,6 +772,63 @@ def accessible_objects(user_or_attribute, operation, options = {})
       end
     end
 
+    # Version of accessible_objects which only returns objects that are
+    # ancestors of a specified root object or the object itself
+    def accessible_ancestor_objects(user_or_attribute, operation, root_object, options = {})
+      # If the root_object is a generic PM::Object, convert it the appropriate storage adapter Object
+      root_object_stored_pe = root_object.try(:stored_pe) || root_object
+
+      # The final set of accessible objects must be ancestors of the root_object
+      ancestor_objects = root_object_stored_pe.ancestors(type: class_for_type('object')) + [root_object_stored_pe]
+
+      # Short-circuit and return all ancestors (minus prohibitions) if the user_or_attribute is authorized on the root node
+      if is_privilege?(user_or_attribute, operation, root_object_stored_pe)
+        apply_include_condition!(ancestor_objects, options[:key], options[:includes])
+
+        if !options[:ignore_prohibitions] && prohibition = prohibition_for(operation)
+          prohibited_ancestor_objects = accessible_ancestor_objects(
+            user_or_attribute,
+            prohibition,
+            root_object_stored_pe,
+            options.merge(ignore_prohibitions: true)
```

Comment:
Don't we _not_ want to ignore prohibitions here?

Reply:
This call uses the prohibition, if one exists. We don't want to continue the recursion from "~read" to "~~read".